src/documents/posts/2016-07-15-self-updating-go-binaries-with-go-selfupdate.md

```diff
@@ -0,0 +1,126 @@
+---
+title: Self-Updating Go Binaries with go-selfupdate
+author: Bob Zoller
+layout: post
+disqus:
+  shortname: goodeggsbytes
+  url: "http://bites.goodeggs.com/post/self-updating-go-binaries-with-go-selfupdate"
+---
+
+The main developer interface to our bespoke PaaS (called Ranch) is a command line binary written in Go.  At first, a Homebrew recipe was plenty convienent to get it onto my early adopter's OSX laptops.  Once I had a few more users and some Linux hosts (CI/CD), however, I wanted to create a built-in update command.
+
+In this post I'll cover how I built the update command, added it to my pre-existing [goxc](https://github.com/laher/goxc)-based build & release flow, and touch on a few improvements I'd like to make down the road.
+
+<!-- more -->
+
+I experimented with a few OSS libraries and considered using the [equinox.io](https://equinox.io/) service, but ended up settling on a simple library called [go-selfupdate](https://github.com/sanbornm/go-selfupdate).
+
+When I say simple I mean simple: go-selfupdate exposes one public API: [`BackgroundRun`](https://github.com/sanbornm/go-selfupdate/blob/master/selfupdate/selfupdate.go#L98).  This method checks a remote HTTP server for a new version, downloads a binary patch if one was found, and applies it in-place.  `BackgroundRun` is meant to be called as a go subroutine, but I'm not a fan of CLIs that update this way.  Instead, I call it inline as part of an explicit `update` command.
+
+I hit one snag along the way that caused me to fork go-selfupdate: it uses an internally-managed TTL and will only check for updates once every 24 hours.  Since I wanted my users to grab updates explicitly, I patched in a `ForceCheck` option and set that to `true`:
```

Comment:
Did you offer a PR to the original project? Seems like a useful addition to lots of gophers out there